### PR TITLE
root: add --registry-mirror flag

### LINKED ISSUE #406: does kaniko support docker registry mirrors ?
dear maintainers

i use kaniko to build docker image at China mainland，when it download base image, it use docker official registry, the GFW will reset this connection, so, does kaniko support docker registry mirrors ?

thank u

### PR DESCRIPTION
**Pull request by greut**
Fixes #406.

**Description**

**Submitter Checklist**

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [unit tests](../DEVELOPMENT.md#creating-a-pr)
- [x] Adds integration tests if needed.

_See [the contribution guide](../CONTRIBUTING.md) for more details._

**Reviewer Notes**

- [ ] The code flow looks good. 
- [ ] Unit tests and or integration tests added.

**Release Notes**

Describe any changes here so maintainer can include it in the release notes, or delete this block.

```
- --registry-mirror <url> to override the default docker.io registry.
```

### CODE REVIEW

**Comment by greut** [comment]
Who could give me hand for the integration tests?

**Comment by tejal29** [comment]
@greut currently our integration tests are not sofisticated :)
you can add a test case here to use the flag `--registry-mirror` https://github.com/GoogleContainerTools/kaniko/blob/master/integration/integration_test.go#L169

**Comment by m86194** [comment]
We need this.  What is necessary for this to move on?

**Comment by cvgw** [comment]
Integration tests failure
```
Building kaniko image
Building cache warmer image
Building onbuild base image
Pushing onbuild base image
Building hardlink base image
Pushing hardlink base image
Building kaniko image using registry-mirror
Building kaniko image using registry-mirror failed: exit status 125unknown flag: --registry-mirror
See 'docker build --help'.
```

**Comment by greut** [comment]
@cvgw Indeed, I didn't take the proper time to understand how the integration test where meant to be written. Let me know if makes more sense now.

File: integration/integration_test.go [hunk 1]
```diff
@@ -183,7 +184,6 @@ func TestRun(t *testing.T) {
 
 func TestGitBuildcontext(t *testing.T) {
 	repo := "github.com/GoogleContainerTools/kaniko"
-	dockerfile := "integration/dockerfiles/Dockerfile_test_run_2"
```
Comment (cvgw) [inline integration/integration_test.go]: This causing the integration tests to fail as the references to `dockerfile` have not been updated 

```
# github.com/GoogleContainerTools/kaniko/integration [github.com/GoogleContainerTools/kaniko/integration.test]
./integration_test.go:193:10: undefined: dockerfile
./integration_test.go:206:10: undefined: dockerfile
./integration_test.go:248:10: undefined: dockerfile
./integration_test.go:286:2: undefined: dockerfile
./integration_test.go:291:10: undefined: dockerfile
```

**Comment by tejal29** [comment]
Restarting the kokoro run.

**Comment by greut** [comment]
@tejal29 sorry to confuse the cla bot a bit more.

**Comment by tejal29** [comment]
> @tejal29 sorry to confuse the cla bot a bit more.

no worries @greut I will take this home!

File: integration/integration_test.go [hunk 2]
```diff
@@ -260,6 +260,50 @@ func TestGitBuildcontext(t *testing.T) {
...
+	}
+
+	// Build with kaniko
+	kanikoImage := GetKanikoImage(config.imageRepo, "Dockerfile_registry_mirror")
+	dockerRunFlags := []string{"run", "--net=host"}
+	dockerRunFlags = addServiceAccountFlags(dockerRunFlags, config.serviceAccount)
+	dockerRunFlags = append(dockerRunFlags, ExecutorImage,
+		"-f", dockerfile,
+		"-d", kanikoImage,
+		"-c", fmt.Sprintf("git://%s", repo))
```
Comment (samos123) [inline integration/integration_test.go]: why isn't this test setting --registry-mirror flag?

File: pkg/util/image_util.go [hunk 3]
```diff
@@ -107,11 +107,25 @@ func remoteImage(image string, opts *config.KanikoOptions) (v1.Image, error) {
 	}
 
 	registryName := ref.Context().RegistryStr()
+	var newReg name.Registry
+
+	if opts.RegistryMirror != "" && registryName == name.DefaultRegistry {
```
Comment (samos123) [inline pkg/util/image_util.go]: this seems to break the following situation:
DefaultRegistry: index.docker.io
mirror-registry unset
Dockerfile starts with from localhost:5000/xxx

localhost:5000 does not require authentication, however following error is seen: Error while retrieving image from cache: localhost:5000/kaniko-test/onbuild-base:latest UNAUTHORIZED: authentication required; [map[Action:pull Class: Name:kaniko-test/onbuild-base Type:repository]] 

See here for more info: https://travis-ci.org/GoogleContainerTools/kaniko/builds/644019948#L3547

File: integration/integration_test.go (hunk 2 above)
Comment (tejal29) [reply integration/integration_test.go]: it should. Let me fix that.

File: pkg/util/image_util.go (hunk 3 above)
Comment (greut) [reply pkg/util/image_util.go]: @samos123 I've managed to reproduce this locally and fix it. Thanks.

**Comment by samos123** [comment]
Kokoro failure is safe to ignore since Kokoro doesn't use the pull request branch for test cases dockerfiles. See error in kokoro here:
```
== RUN   TestBuildViaRegistryMirror
--- FAIL: TestBuildViaRegistryMirror (10.54s)
    integration_test.go:296: Failed to build image gcr.io/kaniko-test/docker-dockerfile_registry_mirror with docker command "[docker build -t gcr.io/kaniko-test/docker-dockerfile_registry_mirror -f integration/dockerfiles/Dockerfile_registry_mirror github.com/GoogleContainerTools/kaniko]": exit status 1 unable to prepare context: unable to evaluate symlinks in Dockerfile path: lstat /tmpfs/tmp/docker-build-git816172650/integration/dockerfiles/Dockerfile_registry_mirror: no such file or directory
    integration_test.go:313: Failed to build image gcr.io/kaniko-test/docker-dockerfile_registry_mirror with kaniko command "[docker run --net=host -v /home/kbuilder/.config/gcloud:/root/.config/gcloud executor-image -f integration/dockerfiles/Dockerfile_registry_mirror -d gcr.io/kaniko-test/kaniko-dockerfile_registry_mirror --registry-mirror us-mirror.gcr.io -c git://github.com/GoogleContainerTools/kaniko]": exit status 1 Enumerating objects: 17206, done.
        Total 17206 (delta 0), reused 0 (delta 0), pack-reused 17206
        Error: error resolving dockerfile path: please provide a valid path to a Dockerfile within the build context with --dockerfile
```

**Comment by darewreck54** [comment]
@greut Is it possible to support multiple registry-mirrors.  If one files, then it would use another one or the default.  This is useful for use cases such as docker hub limitation with API calls.  We could leverage off google mirror cache and if it doesn't have it go to docker hub to pull it into the cache.

**Pull request merged**